src/vcpkg/base/cmd-parser.cpp

```diff
@@ -82,4 +324,692 @@ namespace vcpkg
         }
         m_str.append(line_start, best_break);
     }
+
+    std::vector<std::string> convert_argc_argv_to_arguments(int argc, const CommandLineCharType* const* const argv)
+    {
+        std::vector<std::string> result;
+        // starts at 1 to skip over the program name
+        for (int idx = 1; idx < argc; ++idx)
+        {
+            result.emplace_back(
+#if defined(_WIN32)
+                Strings::to_utf8(argv[idx])
+#else
+                argv[idx]
+#endif // ^^^ !_WIN32
+            );
+        }
+
+        return result;
+    }
+
+    ExpectedL<Unit> replace_response_file_parameters(std::vector<std::string>& inputs,
+                                                     const ILineReader& response_file_source)
+    {
+        auto first = inputs.begin();
+        auto last = inputs.end();
+        while (first != last)
+        {
+            if (first->empty() || first->front() != '@')
+            {
+                ++first;
+                continue;
+            }
+
+            const auto file_name = static_cast<StringView>(*first);
+            auto maybe_response_file_lines = response_file_source.read_lines(file_name.substr(1));
+            if (auto response_file_lines = maybe_response_file_lines.get())
+            {
+                if (response_file_lines->empty())
+                {
+                    first = inputs.erase(first);
+                    last = inputs.end();
+                    continue;
+                }
+
+                // replace the response file name with the first line, and insert the rest
+                auto lines_begin = std::make_move_iterator(response_file_lines->begin());
+                const auto lines_end = std::make_move_iterator(response_file_lines->end());
+                *first = *lines_begin;
+                ++first;
+                ++lines_begin;
+                if (lines_begin != lines_end)
+                {
+                    first = inputs.insert(first, lines_begin, lines_end);
+                    first += response_file_lines->size() - 1;
+                    last = inputs.end();
+                }
+            }
+            else
+            {
+                return std::move(maybe_response_file_lines).error();
+            }
+        }
+
+        return Unit{};
+    }
+
+    bool SwitchName::operator<(const SwitchName& rhs) const { return switch_name < rhs.switch_name; }
+
+    CmdParser::CmdParser()
+        : argument_strings(), argument_strings_lowercase(), argument_parsed(), errors(), options_table()
+    {
+    }
+
+    CmdParser::CmdParser(View<std::string> inputs)
+        : argument_strings(inputs.begin(), inputs.end())
+        , argument_strings_lowercase()
+        , argument_parsed(argument_strings.size(), false)
+        , errors()
+        , options_table()
+    {
+        insert_lowercase_strings(argument_strings_lowercase, argument_strings);
```

Comment:
Because this class is very independent and imperative, it's difficult to see the principles and invariants it follows. The notable exception here being that argument strings are searched as lowercase.

I think it would help clarify the class if additional pre-processing steps were extracted out and done up-front:
1. Compute the subset of arguments that can possibly be options (e.g. double dashes)
2. Pre-parse x- and z- tags

This would also reduce some of the O(N^2) operations going on below, since they happen once per input instead of once per input per switch.

Reply:
1. I am not sure how this helps anything, since we still need to loop through every argument that could possibly be an option on each call that does that.
2. Ditto I'm not sure what this achieves.

Can you be more specific about what you're trying to achieve here?

>This would also reduce some of the O(N^2) operations going on below, since they happen once per input instead of once per input per switch.

To be clear, there is no complexity difference here. The old behavior was arguments * switches, the new behavior is still arguments * switches. It just takes the opposite cut. That is, the old behavior is:

```
foreach (argument in arguments) {
    foreach (switch in switches) {
}}
```

The new behavior is:

```
foreach (switch in switches) {
    foreach (argument in arguments {
}}
```

I considered adding a separate integer that tracked the "frontier" to start at before which we know all arguments are consumed which might speed this up in some cases but ultimately decided that the extra complexity to do that was not worth it.